Arrowgene.Ddon.GameServer/Handler/CraftStartEquipGradeUpHandler.cs

```diff
@@ -0,0 +1,120 @@
+            CDataMDataCraftGradeupRecipe json_data = Server.AssetRepository.CraftingGradeUpRecipesAsset
+                .SelectMany(recipes => recipes.RecipeList)
+                .Where(recipe => recipe.ItemID == equipItemID)
+                .First();
+
+            // Define local variables for calculations
+            uint gearupgradeID = json_data.GradeupItemID;
+            uint goldRequired = json_data.Cost;
+            uint nextGrade = json_data.Unk0; // This might be Unk0 in the JSON but is probably in the DB or something.
+            uint currentTotalEquipPoint = 0; // Equip Points are probably handled elsewhere, since its not in the JSON or Request.
+            uint addEquipPoint = 350;   
+            uint test = json_data.RecipeID;      
+
+            // TODO we need to implement Pawn craft levels since that affects the points that get added
+
+
+            S2CItemUpdateCharacterItemNtc updateCharacterItemNtc = new S2CItemUpdateCharacterItemNtc();
+            updateCharacterItemNtc.UpdateType = 0;
+
+
+            // Remove crafting materials
+            foreach (var craftMaterial in packet.Structure.CraftMaterialList)
+            {
+                try
+                {
+                    List<CDataItemUpdateResult> updateResults = _itemManager.ConsumeItemByUIdFromMultipleStorages(Server, client.Character, STORAGE_TYPES, craftMaterial.ItemUId, craftMaterial.ItemNum);
+                    updateCharacterItemNtc.UpdateItemList.AddRange(updateResults);
+                }
+                catch (NotEnoughItemsException e)
+                {
+                    Logger.Exception(e);
+                    client.Send(new S2CCraftStartEquipGradeUpRes()
+                    {
+                        Result = 1
+                    });
+                    return;
+                }
+            }
+
+
+            // Subtract less Gold if supportpawn is used.
+            if(packet.Structure.CraftSupportPawnIDList.Count > 0)
+            {
+                goldRequired = (uint)(goldRequired*0.95);
+            }
+
+            // Substract Gold based on JSON cost.
+            CDataWalletPoint wallet = client.Character.WalletPointList.Where(wp => wp.Type == WalletType.Gold).Single();
+            wallet.Value = (uint)Math.Max(0, (int)wallet.Value - (int)goldRequired);
+            Database.UpdateWalletPoint(client.Character.CharacterId, wallet);
+            updateCharacterItemNtc.UpdateWalletList.Add(new CDataUpdateWalletPoint()
+            {
+                Type = WalletType.Gold,
+                AddPoint = (int)-goldRequired,
+                Value = wallet.Value
+            });
+
+            // Exchange upgraded items
+            List<CDataItemUpdateResult> AddItemResult = _itemManager.AddItem(Server, client.Character, false, gearupgradeID, 1 * 1);
+            List<CDataItemUpdateResult> RemoveItemResult = _itemManager.ConsumeItemByUIdFromMultipleStorages(Server, client.Character, STORAGE_TYPES, equipItemUID, 1);
```

Comment:
Do it the opposite way, remove the item first and then add the item, in the same storage and slot. This way well avoid errors trying to add an item in a full storage